examples/cfq/models.py

```diff
@@ -91,18 +100,99 @@ def apply(self,
     return attention  # Shape: <float32>[batch_size, seq_len]
 
 
+class MultilayerLSTM(nn.Module):
+  "LSTM cell with multiple layers"
+
+  def apply(self,
+        num_layers,
+        horizontal_dropout_masks,
+        vertical_dropout_masks,
+        input: jnp.array,
+        previous_states: List):
+    """
+    Args
+      num_layers: number of layers
+      horizontal_dropout_masks: dropout masks for each layer (the same dropout
+        mask is used at each time step and applied on the hidden state that is
+        fed into the cell to the right) [num_layers, batch_size, hidden_size]
+      vertical_dropout_masks: dropout masks between layers, same masks reused
+        accross time steps, applied on the hidden state that is fed to the upper
+        layer (used only between layers, not on final output). The dropout is
+        the one used here https://arxiv.org/abs/1512.05287 except it's not
+        applied on the first layer inputs and final outputs to be consistent
+        with the implementation in `recurrent.py`
+        shape [num_layers-1, batch_size, hidden_size]
+      input: input given to the first LSTM layer [batch_size, input_size]
+      previous_states: list of (c,h) for each layer
+        shape [num_layers, batch_size, 2*hidden_size]
+    """
+    cell_input = input
+    states = []
+    final_output = None
+    for layer_idx in range(num_layers):
+      lstm_name = 'lstm_layer'+str(layer_idx)
+      cell = nn.LSTMCell.shared(name=lstm_name)
+      c,h = previous_states[layer_idx]
+      # apply dropout to h
+      if horizontal_dropout_masks[layer_idx] is not None:
+        h = h * horizontal_dropout_masks[layer_idx]
+      # apply dropout to hidden state from lower layer
+      if layer_idx!=0 and vertical_dropout_masks[layer_idx-1] is not None:
+        cell_input = cell_input * vertical_dropout_masks[layer_idx-1]
+      state, output = cell((c,h),cell_input)
+      states.append(state)
+      cell_input = output
+      final_output = output
+    return states, final_output
+
+
 class Decoder(nn.Module):
   """LSTM decoder."""
 
+  def create_dropout_masks(self,
+                           num_masks: int,
+                           shape: Tuple,
+                           dropout_rate: float):
+    if dropout_rate==0:
+      return [None] * num_masks
+    masks = []
+    for i in range(0, num_masks):
+      # should I use a different rng for each mask?
```

Comment:
Did you ask it on FLAXers? What did they reply?

Reply:
I thought it might not be worth asking - I am thinking if I am using the same key I will be getting the same random distribution, which is not what I want. I guess I can put it in a colab and test it out but it feels like a safe assumption